app/authentication/jwt_decoder.py

```diff
@@ -37,22 +69,81 @@ def decode_jwt_token(self, token):
     def decode_signed_jwt_token(self, signed_token):
         try:
             if signed_token:
+                logging.debug("Decoding signed JWT " + self.__to_str(signed_token))
+                self._check_headers(signed_token)
+                self._check_payload(signed_token)
                 token = jwt.decode(signed_token, self.rrm_public_key, algorithms=['RS256'])
+                self._check_token(token)
+                if not token:
+                    raise InvalidTokenException("Missing Payload")
                 return token
             else:
                 raise NoTokenException("JWT Missing")
-        except jwt.DecodeError as e:
+        except (jwt.DecodeError,
+                jwt.exceptions.InvalidAlgorithmError,
+                jwt.exceptions.ExpiredSignatureError,
+                jwt.exceptions.InvalidIssuedAtError) as e:
             raise InvalidTokenException(repr(e))
-        except jwt.exceptions.InvalidAlgorithmError as e:
+
+    def _check_headers(self, signed_token):
+        token = self.__to_str(signed_token)
+        header_data = token.split('.', 1)[0]
+        try:
+            headers = self._base64_decode(header_data)
+            self.__check_for_duplicates(headers)
+        except UnicodeDecodeError:
+            raise InvalidTokenException("Corrupted Header")
+        except ValueError as e:
             raise InvalidTokenException(repr(e))
 
+        header = jwt.get_unverified_header(token)
+        if not header:
+            raise InvalidTokenException("Missing Headers")
+        if not header.get('typ'):
+            raise InvalidTokenException("Missing Type")
+        if not header.get('alg'):
+            raise InvalidTokenException("Missing Algorithm")
+        if not header.get('kid'):
+            raise InvalidTokenException("Missing kid")
+        if "JWT" != header.get('typ').upper():
+            raise InvalidTokenException("Invalid Type")
+        if "RS256" != header.get('alg').upper():
+            raise InvalidTokenException("Invalid Algorithm")
+        if "EDCRRM" != header.get('kid').upper():
+            raise InvalidTokenException("Invalid Key Identifier")
+
+    def __check_for_duplicates(self, headers):
+        h = self.__to_str(headers)
+        if h.count("typ") > 1:
```

Comment:
There is a risk with this approach that a header containing this value not as a 'key' will fail such as {"key1":"type1"} would hit this trigger because the value contains 'typ'. Something to consider.